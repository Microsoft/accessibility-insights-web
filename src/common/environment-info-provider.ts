// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { toolName } from '../content/strings/application';
import { ToolData } from './types/store-data/unified-data-interface';

export type EnvironmentInfo = {
    extensionVersion: string;
    browserSpec: string;
    axeCoreVersion: string;
};

export class EnvironmentInfoProvider {
    constructor(
        private extensionVersion: string,
        private browserSpec: string,
        private axeVersion: string,
    ) {}
    public getEnvironmentInfo(): EnvironmentInfo {
        return {
            extensionVersion: this.extensionVersion,
            browserSpec: this.browserSpec,
            axeCoreVersion: this.axeVersion,
        };
    }

    public getToolData(): ToolData {
        return {
            scanEngineProperties: {
                name: 'axe-core',
                version: this.axeVersion,
            },
            applicationProperties: {
                name: toolName,
                version: this.extensionVersion,
            },
        };
    }
}
