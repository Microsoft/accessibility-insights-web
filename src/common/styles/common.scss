// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
@import 'high-contrast-override.scss';
@import 'colors.scss';
@import 'fonts.scss';

$focus-ring-color: $neutral-100;

@mixin set-focus-ring-styles($property: 'border') {
    #{$property}: 1px solid $focus-ring-color;
    @include ms-high-contrast-override {
        #{$property}: 1px dashed windowtext !important;
    }
}

.ms-Fabric.is-focusVisible .ms-Nav-group .ms-nav-linkButton:focus::after {
    @include set-focus-ring-styles('border');
}

button::after {
    content: '';
    position: absolute;
}

.ms-Button.ms-Button--primary:not(.is-disabled) {
    background: $communication-primary;

    &:hover {
        background: $communication-shade-10;
    }
}

.ms-fontColor-neutralPrimary,
.ms-fontColor-neutralPrimary--hover:hover {
    color: $neutral-100 !important;
}

.ms-Fabric {
    color: $neutral-100;
}

.generic-toggle-component {
    margin-bottom: 4vh;
    .toggle-container {
        margin-bottom: 1vh;
        .toggle-name {
            display: inline-block;
            font-size: 17px;
            font-weight: bold;
        }

        .toggle {
            float: right;
        }
    }
}

.high-contrast-theme {
    .insights-link {
        &:hover {
            text-decoration: underline;
        }
    }
    .is-selected {
        .status-icon {
            color: $high-contrast-icon-color !important;
        }
    }
}

.insights-link {
    // marking important to override office fabric style
    color: $link !important;
    text-decoration: none;

    &:hover {
        color: $link-hover !important;
        text-decoration: underline;
    }
}

.ms-Spinner {
    .ms-Spinner-circle {
        border-top-color: $spinner-text;
    }

    .ms-Spinner-label {
        color: $spinner-text;
    }
}

$normalTitleFontWeight: 300 !default;

$pivotItemLeftBorderWidth: 3px !default;
$pivotItemBorderStyle: solid !default;

$detailsViewHeaderBarHeight: 40px !default;
$detailsViewNavPivotsHeight: 41px !default;
$detailsViewCommandBarHeight: 40px !default;
$detailsViewCommandBarBorderHeight: 1px !default;
$detailsViewLeftNavWidth: 200px !default;
$detailsViewTotalHeaderHeight: (#{$detailsViewCommandBarBorderHeight} + #{$detailsViewCommandBarHeight} + #{$detailsViewHeaderBarHeight});

$fastPassRightPanelMarginRight: 24px !default;
$fastPassRightPanelMarginLeft: 24px !default;
$fastPassRightPanelMarginTop: 24px !default;

@mixin ellipsedText {
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
}
