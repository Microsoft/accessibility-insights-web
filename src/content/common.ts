// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as react from 'react';

import { link } from './link';
import { ContentCreator } from '../views/content/content-page';
import { GuidanceTitle as gt } from '../views/content/guidance-title';

export const create = ContentCreator(link);
export const React = react;
export const GuidanceTitle = gt;
