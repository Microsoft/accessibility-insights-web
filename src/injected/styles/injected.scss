// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
@import '../../../node_modules/office-ui-fabric-react/dist/sass/Icon.Definitions';
@import '../../common/styles/colors.scss';

#accessibility-insights-root-container {
    visibility: hidden !important;
}

#accessibility-insights-root-container,
#insights-shadow-container {
    @import '../../common/styles/common.scss';
    @import '../../common/components/toast.scss';

    @mixin max-z-index-1 {
        z-index: 2147483646 !important;
    }

    @mixin max-z-index {
        z-index: 2147483647 !important;
    }

    @include max-z-index();

    position: absolute !important;
    top: 0px !important;
    left: 0px !important;
    padding: 0px !important;
    margin: 0px !important;
    width: 100% !important;
    height: 100% !important;

    .insights-dialog-container {
        @import '../../../node_modules/office-ui-fabric-react/dist/sass/Fabric.scss';

        @include max-z-index();
    }

    .insights-dialog-main-override-shadow {
        position: fixed;
        top: 0;
        left: 0;
        right: 0;
        bottom: 0;
        background-color: $neutral-alpha-60;
        @include max-z-index();
        font-family: $fontFamily;
    }

    div.insights-dialog-main-container {
        width: 25%;
        max-width: 25%;
        user-select: text;
        min-width: 520px;
    }

    div.insights-file-issue-details-dialog-container {
        max-width: 370px;

        .ms-Dialog-content p {
            margin-block-start: 4px;
        }

        .ms-Dialog-actions {
            margin-top: 24px;
        }
    }

    #insights-inspect-selector-layout {
        z-index: 999999999 !important;
    }

    .insights-modal {
        overflow: hidden !important;
    }

    .insights-dialog-main-override .insights-dialog-main-container .ms-Dialog-title {
        padding-left: 19px;
    }

    .insights-dialog-main-override-shadow {
        .ms-Dialog-title {
            font-family: $fontFamily;
        }

        .ms-button-flex-container {
            display: flex;
        }
    }

    .insights-dialog-main-override .insights-dialog-main-container .ms-Dialog-inner {
        padding: 0 19px 20px !important;
    }

    .insights-dialog-main-override .insights-dialog-button-inspect i {
        color: $communication-shade-10 !important;
    }

    .insights-dialog-main-override .insights-dialog-button-inspect.is-disabled i {
        color: $neutral-30 !important;
    }

    .insights-dialog-main-override-shadow .insights-dialog-button-inspect i {
        color: $communication-shade-10 !important;
        font-style: normal;
        font-family: FabricMDL2Icons;
    }

    .insights-dialog-main-override-shadow .insights-dialog-button-inspect.is-disabled i {
        color: $neutral-30 !important;
        font-style: normal;
        font-family: FabricMDL2Icons;
    }

    .insights-dialog-main-override .create-bug-button,
    .insights-dialog-main-override .copy-issue-details-button {
        margin-left: 8px !important;
    }

    .insights-dialog-main-override .insights-dialog-button-create-bug i {
        color: $communication-shade-10 !important;
    }

    .insights-dialog-main-override-shadow .insights-dialog-button-create-bug {
        margin-left: 8px !important;
    }

    .insights-dialog-main-override-shadow .insights-dialog-button-create-bug i {
        color: $communication-shade-10 !important;
        font-style: normal;
        font-family: FabricMDL2Icons;
    }

    .insights-dialog-main-override .insights-dialog-button-left {
        float: left !important;
    }

    .insights-dialog-main-override .insights-dialog-button-right {
        float: right !important;
        text-align: right !important;
    }

    .insights-dialog-main-override .insights-dialog-footer {
        text-align: center !important;
        margin-top: 5px !important;
    }

    .insights-dialog-main-override-shadow .insights-dialog-footer {
        text-align: center !important;
        float: left !important;
        margin-top: 5px !important;
        width: 50%;
        font-size: 14px;
    }

    .insights-dialog-main-override ul,
    .insights-dialog-main-override-shadow ul {
        margin-top: 3px !important;
    }

    .insights-dialog-main-override .insights-dialog-rule-link,
    .insights-dialog-main-override-shadow .insights-dialog-rule-link {
        margin-top: 5px !important;
        margin-bottom: 6px !important;
        padding: 0 !important;

        .ms-Link {
            color: $communication-shade-10;
        }
    }

    .insights-dialog-main-override .insights-dialog-rule-container,
    .insights-dialog-main-override-shadow .insights-dialog-rule-container {
        margin-bottom: 12px !important;
    }

    .insights-dialog-main-override-shadow .insights-dialog-container {
        margin-bottom: 12px !important;
        position: relative;
        background-color: $neutral-2;
        margin: auto;
        padding: 16px 20px;
        border: 1px solid $neutral-60;
        width: 35em;
        min-width: 27%;
        text-align: left;
        box-shadow: 0 4px 8px 0 $neutral-alpha-20, 0 6px 20px 0 $neutral-alpha-20;
        animation-name: animatetop;
        animation-duration: 0.4s;
        align-items: center;
        top: 30%;

        .insights-dialog-header {
            margin-top: 15px;

            .ms-Dialog-title {
                display: inline;
                font-size: 21px;
                font-weight: 100;
            }

            .ms-Dialog-topButton {
                button {
                    background-color: transparent;
                    border: none;
                    font-size: 16px;
                    cursor: pointer;

                    &:hover {
                        color: $communication-shade-30;
                    }
                }

                display: inline;
                float: right;

                i {
                    font-style: normal;
                    font-family: FabricMDL2Icons;
                }
            }
        }
    }

    .insights-dialog-main-override-shadow .ms-Grid .ms-Grid-row {
        button {
            position: relative;
            font-family: $fontFamily;
            -webkit-font-smoothing: antialiased;
            font-size: 14px;
            font-weight: 400;
            box-sizing: border-box;
            display: inline-block;
            text-align: center;
            cursor: pointer;
            vertical-align: top;
            padding: 0px 16px;
            min-width: 80px;
            height: 32px;
            background-color: $communication-primary;
            color: $neutral-0;
            user-select: none;
            outline: transparent;
            border-width: 1px;
            border-style: solid;
            border-color: transparent;
            border-image: initial;
            text-decoration: none;
            border-radius: 0px;

            &:focus {
                outline: -webkit-focus-ring-color auto 5px;
            }
        }

        button:disabled {
            background-color: $neutral-4;
            color: $neutral-30;
            pointer-events: none;
        }

        .ms-Grid-col.insights-dialog-button-left {
            padding-left: 0px;
            float: left;
        }

        .ms-Grid-col.insights-dialog-button-right {
            padding-right: 0px;
            float: right;
            text-align: right;
        }
    }

    .insights-dialog-main-override .insights-dialog-rule-container i {
        color: $negative-outcome !important;
        float: left !important;
        margin-right: 1% !important;
        margin-top: 3px !important;
    }

    .insights-dialog-main-override-shadow .insights-dialog-rule-container i {
        color: $negative-outcome !important;
        float: left !important;
        margin-right: 1% !important;
        margin-top: 2px !important;
        font-style: normal;
        font-family: FabricMDL2Icons;
    }

    .insights-dialog-main-override .insights-dialog-target-container {
        margin-bottom: 16px !important;
        font-size: 14px;
    }

    .insights-dialog-main-override .insights-dialog-target-button-container {
        margin-top: 4px !important;
    }

    .insights-dialog-main-override-shadow .insights-dialog-target-button-container {
        margin-top: 4px;

        button {
            position: relative;
            font-family: $fontFamily;
            -webkit-font-smoothing: antialiased;
            font-size: 14px;
            font-weight: 400;
            box-sizing: border-box;
            display: inline-block;
            text-align: center;
            cursor: pointer;
            vertical-align: top;
            padding: 0px 16px;
            min-width: 80px;
            height: 32px;
            background-color: $neutral-4;
            color: $neutral-80;
            user-select: none;
            outline: transparent;
            border-width: 1px;
            border-style: solid;
            border-color: transparent;
            border-image: initial;
            text-decoration: none;
            border-radius: 0px;

            &:focus {
                outline: -webkit-focus-ring-color auto 5px;
            }
        }

        button:disabled {
            background-color: $neutral-4;
            color: $neutral-30;
            pointer-events: none;
        }

        .ms-Button-label {
            margin-left: 4px;
            line-height: 100%;
            font-weight: 600;
        }
    }

    .insights-dialog-main-override .insights-dialog-inspect-disabled {
        margin-top: 4px !important;
        font-size: 12px !important;
    }

    .insights-dialog-main-override .create-bug-button-help {
        margin-top: 4px !important;
        margin-bottom: 8px !important;
        font-size: 12px !important;
        color: $negative-outcome;
        font-weight: $fontWeightSemiBold;
    }

    .insights-dialog-main-override .insights-dialog-fix-instruction-container {
        margin-bottom: 20px !important;
        font-size: 14px;
    }

    .insights-dialog-main-override .insights-fix-instruction-list {
        padding-left: 5% !important;
        margin: 0 !important;
    }

    .insights-dialog-main-override .insights-fix-instruction-title {
        font-weight: 600 !important;
    }

    .insights-dialog-main-override-shadow .insights-fix-instruction-title {
        font-weight: 600 !important;
        margin-top: 15px;
    }

    .insights-dialog-main-override .insights-dialog-instance-selector {
        padding-left: 2% !important;
    }

    .insights-highlight-container,
    .insights-highlight-container svg {
        position: absolute !important;
        @include max-z-index-1();
        top: 0 !important;
        left: 0 !important;
        overflow: visible !important;
        pointer-events: none !important;
        visibility: visible !important;
    }

    .insights-highlight-container .insights-highlight-box {
        visibility: visible !important;
        position: absolute !important;
        overflow: hidden !important;
        pointer-events: none !important;
    }

    .insights-highlight-container .insights-highlight-box .insights-highlight-text {
        cursor: default !important;
        pointer-events: all !important;
        font-size: 1em !important;
        padding: 0.3em !important;
        float: right !important;
        position: relative !important;
        text-align: center !important;
    }

    :host {
        @include max-z-index-1();
        top: 0 !important;
        left: 0 !important;
        width: 0 !important;
        height: 0 !important;
        visibility: hidden !important;
        padding: 0 0 0 0 !important;
        margin: 0 0 0 0 !important;
    }

    .insights-grey-scale-container {
        filter: grayscale(100%) !important;
    }

    .insights-pseudo-selector-style-container {
        *::before,
        *::after {
            border: 1px solid red;
        }
    }
}

// We need the below styling only when not using shadow dom
#accessibility-insights-root-container > .insights-dialog-container {
    position: absolute !important;
    top: 0px !important;
    left: 0px !important;
    padding: 0px !important;
    margin: 0px !important;
}
