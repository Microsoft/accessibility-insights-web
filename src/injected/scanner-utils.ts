// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { autobind } from '@uifabric/utilities';

import { createDefaultLogger } from '../common/logging/default-logger';
import { Logger } from '../common/logging/logger';
import { scan as scanRunner, ScanOptions } from '../scanner/exposed-apis';
import { RuleResult, ScanResults } from '../scanner/iruleresults';
import { HyperlinkDefinition } from '../views/content/content-page';

declare var axe: any;

export interface DecoratedAxeNodeResult {
    any: FormattedCheckResult[];
    none: FormattedCheckResult[];
    all: FormattedCheckResult[];
    status: boolean;
    ruleId: string;
    failureSummary: string;
    selector: string;
    html: string;
    help: string;
    id: string;
    guidanceLinks: HyperlinkDefinition[];
    helpUrl: string;
    fingerprint: string;
    snippet: string;
}

// tslint:disable-next-line:interface-name
export interface IHtmlElementAxeResults {
    ruleResults: IDictionaryStringTo<DecoratedAxeNodeResult>;
    isVisible: boolean;
    propertyBag?: any;
    target: string[];
}

export class ScannerUtils {
    public constructor(
        private scanner: typeof scanRunner,
        private generateUID?: () => string,
        private logger: Logger = createDefaultLogger(),
    ) {}

    public scan(options: ScanOptions, callback: (results: ScanResults) => void): void {
        this.scanner(
            options,
            axeResults => {
                callback(axeResults);
            },
            err => {
                this.logger.log(`failed to scan with error - ${err}`);
            },
        );
    }

    public getUniqueSelector(element: HTMLElement): string {
        /* axe team's suggestion to how to use getSelector */
        axe._tree = axe.utils.getFlattenedTree(document.documentElement);
        axe._selectorData = axe.utils.getSelectorData(axe._tree);
        const selector = axe.utils.getSelector(element);
        axe._tree = undefined;
        axe._selectorData = undefined;
        return selector;
    }

    @autobind
    public getIncompleteInstances(results: ScanResults): IDictionaryStringTo<IHtmlElementAxeResults> {
        const resultsMap: IDictionaryStringTo<IHtmlElementAxeResults> = {};
        this.addIncompletesToDictionary(resultsMap, results.incomplete);
        return resultsMap;
    }

    @autobind
    public getFailingInstances(results: ScanResults): IDictionaryStringTo<IHtmlElementAxeResults> {
        const resultsMap: IDictionaryStringTo<IHtmlElementAxeResults> = {};
        this.addFailuresToDictionary(resultsMap, results.violations);
        return resultsMap;
    }

    @autobind
    public getPassingInstances(results: ScanResults): IDictionaryStringTo<IHtmlElementAxeResults> {
        const resultsMap: IDictionaryStringTo<IHtmlElementAxeResults> = {};
        this.addPassesToDictionary(resultsMap, results.passes);
        return resultsMap;
    }

    @autobind
    public getAllCompletedInstances(results: ScanResults): IDictionaryStringTo<IHtmlElementAxeResults> {
        const resultsMap: IDictionaryStringTo<IHtmlElementAxeResults> = {};
        this.addPassesToDictionary(resultsMap, results.passes);
        this.addFailuresToDictionary(resultsMap, results.violations);
        return resultsMap;
    }

    @autobind
    public getFailingOrPassingInstances(results: ScanResults): IDictionaryStringTo<IHtmlElementAxeResults> {
        const resultsMap: IDictionaryStringTo<IHtmlElementAxeResults> = {};
        this.addFailuresToDictionary(resultsMap, results.violations);
        if (Object.keys(resultsMap).length === 0) {
            this.addPassesToDictionary(resultsMap, results.passes);
        }
        return resultsMap;
    }

    private addPassesToDictionary(dictionary: IDictionaryStringTo<IHtmlElementAxeResults>, axeRules: RuleResult[]): void {
        this.addResultstoDictionary(dictionary, axeRules, true);
    }

    private addIncompletesToDictionary(dictionary: IDictionaryStringTo<IHtmlElementAxeResults>, axeRules: RuleResult[]): void {
        this.addResultstoDictionary(dictionary, axeRules, undefined);
    }

    private addFailuresToDictionary(dictionary: IDictionaryStringTo<IHtmlElementAxeResults>, axeRules: RuleResult[]): void {
        this.addResultstoDictionary(dictionary, axeRules, false);
    }

    private addResultstoDictionary(dictionary: IDictionaryStringTo<IHtmlElementAxeResults>, axeRules: RuleResult[], status: boolean): void {
        axeRules.forEach(ruleResult => {
            ruleResult.nodes.forEach(node => {
                const selectorKey = node.target.join(';');
                node.instanceId = this.generateUID ? this.generateUID() : null;

                const elementResult = dictionary[selectorKey] || {
                    target: node.target,
                    ruleResults: {},
                    isVisible: true,
                };

                dictionary[selectorKey] = elementResult;
                elementResult.ruleResults[ruleResult.id] = {
                    any: node.any,
                    all: node.all,
                    none: node.none,
                    status: status,
                    ruleId: ruleResult.id,
                    help: ruleResult.help,
                    failureSummary: node.failureSummary,
                    html: node.html,
                    selector: selectorKey,
                    id: node.instanceId,
                    guidanceLinks: ruleResult.guidanceLinks,
                    helpUrl: ruleResult.helpUrl,
                    fingerprint: ScannerUtils.getFingerprint(node, ruleResult),
                    snippet: node.snippet || node.html,
                };
            });
        });
    }

    public static getFingerprint(node: AxeNodeResult, rule: RuleResult): string {
        return 'fp--' + rule.id + '--' + node.target.join(';') + '--' + node.html;
    }
}
