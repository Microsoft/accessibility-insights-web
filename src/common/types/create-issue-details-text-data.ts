// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { DecoratedAxeNodeResult } from '../../injected/scanner-utils';
import { EnvironmentInfo } from '../environment-info-provider';
import { UnifiedRule } from './store-data/unified-data-interface';

export interface CreateIssueDetailsTextData {
    rule: UnifiedRule; // uses every part of rule in bug filing
    targetApp: {
        name: string;
        url?: string;
    };
    // everything below this line needs to be
    // computed from the unified result
    element: {
        id: string;
        idDisplayName: string;
        shortId: string;
    };
    howToFixSummary: string;
    snippet?: string;
}
