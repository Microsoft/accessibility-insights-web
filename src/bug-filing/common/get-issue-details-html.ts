// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { escape } from 'lodash';
import { EnvironmentInfo } from '../../common/environment-info-provider';
import { CreateIssueDetailsTextData } from '../../common/types/create-issue-details-text-data';
import { title } from '../../content/strings/application';
import { IssueUrlCreationUtils } from './issue-filing-url-string-utils';

const buildBodySection = (headingHtml: string, contentHtml: string): string => {
    return `<div><b>${headingHtml}</b></div><div>${contentHtml}</div><br>`;
};

const buildIssueDetails = (help: string, helpUrl: string, ruleId: string): string => {
    const helpEscapedForHtml = escape(help);
    const ruleIdEscapedForHtml = escape(ruleId);

    return `${helpEscapedForHtml}<br><a href="${helpUrl}">${ruleIdEscapedForHtml}</a>`;
};

const buildApplicationHtml = (pageTitle: string, pageUrl: string): string => {
    const pageTitleEscapedForHtml = escape(pageTitle);
    const pageUrlEscapedForUrl = encodeURI(pageUrl);
    const pageUrlEscapedForHtml = escape(pageUrl);

    return `<a href="${pageUrlEscapedForUrl}">${pageTitleEscapedForHtml}<br>${pageUrlEscapedForHtml}</a>`;
};

const buildSnippetHtml = (snippet: string): string => {
    const maxSnippetLength = 256;
    let constrainedSnippet = snippet;

    if (snippet.length > maxSnippetLength) {
        constrainedSnippet = snippet.substr(0, maxSnippetLength) + '...';
    }

    return escape(constrainedSnippet);
};

const buildHowToFixHtml = (failureSummary: string): string => {
    return escape(failureSummary)
        .replace(/\n  /g, '<br>- ')
        .replace(/\n /g, '<br> ')
        .replace(/\n/g, '<br>');
};

export const getIssueDetailsHtml = (
    stringUtils: IssueUrlCreationUtils,
    environmentInfo: EnvironmentInfo,
    data: CreateIssueDetailsTextData,
): string => {
    const body =
        buildBodySection('Issue Details', buildIssueDetails(data.ruleResult.help, data.ruleResult.helpUrl, data.ruleResult.ruleId)) +
        buildBodySection('Application', buildApplicationHtml(data.pageTitle, data.pageUrl)) +
        buildBodySection('Element Path', escape(data.ruleResult.selector)) +
        buildBodySection('Snippet', buildSnippetHtml(data.ruleResult.html)) +
        buildBodySection('How to fix', buildHowToFixHtml(data.ruleResult.failureSummary)) +
        buildBodySection(
            'To reproduce',
            `Use <a href="http://aka.ms/AccessibilityInsights">${title}</a> to investigate the issue details`,
        ) +
        buildBodySection('Environment', escape(environmentInfo.browserSpec)) +
        stringUtils.getFooter(environmentInfo);

    return body;
};
