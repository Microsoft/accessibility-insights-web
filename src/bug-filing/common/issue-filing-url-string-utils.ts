// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { endsWith } from 'lodash';

import { EnvironmentInfo } from '../../common/environment-info-provider';
import { title } from '../../content/strings/application';
import { CreateIssueDetailsTextData } from './../../common/types/create-issue-details-text-data';

export class IssueFilingUrlStringUtils {
    public static footer(environmentInfo: EnvironmentInfo): string {
        return (
            `This accessibility issue was found using ${title} ` +
            `${environmentInfo.extensionVersion} (axe-core ${environmentInfo.axeCoreVersion}), ` +
            'a tool that helps find and fix accessibility issues. Get more information & download ' +
            'this tool at http://aka.ms/AccessibilityInsights.'
        );
    }

    public static collapseConsecutiveSpaces(input: string): string {
        return input.replace(/\s+/g, ' ');
    }

    public static markdownEscapeBlock(input: string): string {
        return input
            .split('\n')
            .map(line => `    ${line}`)
            .join('\n');
    }

    public static getSelectorLastPart(selector: string): string {
        let selectorLastPart = selector;
        if (selector.lastIndexOf(' > ') > 0) {
            selectorLastPart = selector.substr(selector.lastIndexOf(' > ') + 3);
        }
        return selectorLastPart;
    }

    public static standardizeTags(data: CreateIssueDetailsTextData): string[] {
        return data.ruleResult.guidanceLinks.map(tag => tag.text.toUpperCase());
    }

    public static appendSuffixToUrl(url: string, suffix: string): string {
        if (endsWith(url, suffix) || endsWith(url, `${suffix}/`)) {
            return url;
        }

        return `${url}/${suffix}/`;
    }
}
