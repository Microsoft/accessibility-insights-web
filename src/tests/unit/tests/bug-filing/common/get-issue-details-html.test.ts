// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { repeat } from 'lodash';
import { IMock, Mock } from 'typemoq';
import { getIssueDetailsHtml } from '../../../../../bug-filing/common/get-issue-details-html';
import { IssueUrlCreationUtils } from '../../../../../bug-filing/common/issue-filing-url-string-utils';
import { EnvironmentInfo } from '../../../../../common/environment-info-provider';
import { CreateIssueDetailsTextData } from '../../../../../common/types/create-issue-details-text-data';
import { DecoratedAxeNodeResult } from '../../../../../injected/scanner-utils';

describe('getIssueDetailsHtml', () => {
    let stringUtilsMock: IMock<IssueUrlCreationUtils>;

    let sampleIssueDetailsData: CreateIssueDetailsTextData;

    const environmentInfo: EnvironmentInfo = {
        extensionVersion: '1.1.1',
        axeCoreVersion: '2.2.2',
        browserSpec: 'test spec',
    };

    const testObject = getIssueDetailsHtml;

    beforeEach(() => {
        stringUtilsMock = Mock.ofType<IssueUrlCreationUtils>();
        stringUtilsMock.setup(utils => utils.getFooter(environmentInfo)).returns(() => 'test footer content');
        sampleIssueDetailsData = {
            pageTitle: 'pageTitle<x>',
            pageUrl: 'pageUrl',
            ruleResult: {
                failureSummary: 'RR-failureSummary',
                guidanceLinks: [{ text: 'WCAG-1.4.1' }, { text: 'wcag-2.8.2' }],
                help: 'RR-help',
                html: 'RR-html',
                ruleId: 'RR-rule-id',
                helpUrl: 'RR help url',
                selector: 'RR-selector<x>',
                snippet: 'RR-snippet   space',
            } as DecoratedAxeNodeResult,
        };
    });

    it('returns issue details text as html', () => {
        const result = testObject(stringUtilsMock.object, environmentInfo, sampleIssueDetailsData);

        expect(result).toMatchSnapshot();
    });

    it('truncates long snippets', () => {
        sampleIssueDetailsData.ruleResult.html = `${repeat('-', 300)}end-of-html`;

        const result = testObject(stringUtilsMock.object, environmentInfo, sampleIssueDetailsData);

        expect(result).not.toContain('end-of-html');
        expect(result).toContain('...');
        expect(result).toMatchSnapshot();
    });
});
