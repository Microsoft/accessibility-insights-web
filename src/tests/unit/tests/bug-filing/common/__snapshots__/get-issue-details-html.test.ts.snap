// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getIssueDetailsHtml returns issue details text as html 1`] = `"<div><b>Issue</b></div><div>RR-help (<a href=\\"RR%20help%20url\\">RR-rule-id</a>)</div><br><div><b>Application</b></div><div><a href=\\"pageUrl\\">pageTitle&lt;x&gt;<br>pageUrl</a></div><br><div><b>Element Path</b></div><div>RR-selector&lt;x&gt;</div><br><div><b>Snippet</b></div><div>RR-html</div><br><div><b>How to fix</b></div><div>RR-failureSummary</div><br><div><b>To reproduce</b></div><div>Use <a href=\\"http://aka.ms/AccessibilityInsights\\">Accessibility Insights for Web</a> to investigate the issue details</div><br><div><b>Environment</b></div><div>test spec</div><br>test footer content"`;

exports[`getIssueDetailsHtml truncates long snippets 1`] = `"<div><b>Issue</b></div><div>RR-help (<a href=\\"RR%20help%20url\\">RR-rule-id</a>)</div><br><div><b>Application</b></div><div><a href=\\"pageUrl\\">pageTitle&lt;x&gt;<br>pageUrl</a></div><br><div><b>Element Path</b></div><div>RR-selector&lt;x&gt;</div><br><div><b>Snippet</b></div><div>----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------...</div><br><div><b>How to fix</b></div><div>RR-failureSummary</div><br><div><b>To reproduce</b></div><div>Use <a href=\\"http://aka.ms/AccessibilityInsights\\">Accessibility Insights for Web</a> to investigate the issue details</div><br><div><b>Environment</b></div><div>test spec</div><br>test footer content"`;
